k2/python/k2/fsa_algo.py

```diff
@@ -50,37 +51,55 @@ def top_sort(fsa: Fsa) -> Fsa:
     for name, value in fsa.named_tensor_attr():
         setattr(sorted_fsa, name, value.index_select(0, arc_map))
     for name, value in fsa.named_non_tensor_attr():
-        setattr(sorted_fsa, name, deepcopy(value))
+        setattr(sorted_fsa, name, value)
     return sorted_fsa
 
 
 def intersect(a_fsa: Fsa, b_fsa: Fsa) -> Fsa:
     '''Compute the intersection of two FSAs.
 
-    Note:
-      Intersection is also known as compose in FST.
-
     Args:
       a_fsa:
         The first input FSA. It can be either a single FSA or a vector of FSAs.
       b_fsa:
         The second input FSA. It can be either a single FSA or a vector of FSAs.
 
+    CAUTION:
+      ``aux_labels`` are discarded since it is not clear how to set the
+      aux_labels of the output FSA.
+
     Returns:
       The result of intersecting a_fsa and b_fsa.
     '''
+    if not a_fsa.arc_labels_sorted:
+        a_fsa = arc_sort(a_fsa)
+
+    if not b_fsa.arc_labels_sorted:
+        b_fsa = arc_sort(b_fsa)
+
     need_arc_map = True
     ragged_arc, a_arc_map, b_arc_map = _k2.intersect(a_fsa.arcs, b_fsa.arcs,
                                                      need_arc_map)
-    a_arc_map = a_arc_map.to(torch.int64)  # required by index_select
-    b_arc_map = b_arc_map.to(torch.int64)
+
+    # Some of entries in a_arc_map and b_arc_map may be -1.
+    # The arc_maps are incremented so that every entry is non-negative.
+    a_arc_map = a_arc_map.to(torch.int64) + 1
```

Comment:
This is OK in the short term, but I think it would be significantly more efficient to create a k2 operation that does something
of the form:
`   c[:] = a[b]`
where c, a and b are tensors, and that sets the element to zero if the index was -1.  Assuming PyTorch doesn't already have an operation like this.